metadata/src/main/java/org/apache/kafka/metadata/MetadataCache.java

```diff
@@ -0,0 +1,233 @@
+            topic.partitions().forEach((partitionId, partition) -> {
+                List<Node> nodes = brokerToNodes.get(partition.leader);
+                if (nodes != null) {
+                    nodes.forEach(node -> {
+                        partitionInfos.add(new PartitionInfo(
+                            topic.name(),
+                            partitionId,
+                            node,
+                            toArray(partition.replicas, brokerToNodes),
+                            toArray(partition.isr, brokerToNodes),
+                            getOfflineReplicas(image, partition).stream()
+                                .map(brokerToNodes::get)
+                                .flatMap(Collection::stream)
+                                .toArray(Node[]::new)
+                        ));
+                    });
+                    if (Topic.isInternal(topic.name())) {
+                        internalTopics.add(topic.name());
+                    }
+                }
+            });
+        });
+
+        Node controllerNode = Optional.ofNullable(brokerToNodes.get(getRandomAliveBroker(image).orElse(-1)))
+            .map(nodes -> nodes.get(0))
+            .orElse(Node.noNode());
+
+        return new Cluster(
+            clusterId,
+            brokerToNodes.values().stream().flatMap(Collection::stream).collect(Collectors.toList()),
+            partitionInfos,
+            Collections.emptySet(),
+            internalTopics,
+            controllerNode
+        );
+    }
+
+    private static Node[] toArray(int[] replicas, Map<Integer, List<Node>> brokerToNodes) {
+        return Arrays.stream(replicas)
+            .mapToObj(brokerToNodes::get)
+            .flatMap(Collection::stream)
+            .toArray(Node[]::new);
+    }
+
+    private static List<Integer> getOfflineReplicas(MetadataImage image, PartitionRegistration partition) {
+        List<Integer> offlineReplicas = new ArrayList<>();
+        for (int brokerId : partition.replicas) {
+            BrokerRegistration broker = image.cluster().broker(brokerId);
+            if (broker == null || isReplicaOffline(partition, broker)) {
+                offlineReplicas.add(brokerId);
+            }
+        }
+        return offlineReplicas;
+    }
+
+    private static boolean isReplicaOffline(PartitionRegistration partition, BrokerRegistration broker) {
+        return broker.fenced() || !broker.hasOnlineDir(partition.directory(broker.id()));
+    }
+
+    private static Optional<Integer> getRandomAliveBroker(MetadataImage image) {
+        List<BrokerMetadata> aliveBrokers = getAliveBrokers(image);
+        if (aliveBrokers.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(aliveBrokers.get(ThreadLocalRandom.current().nextInt(aliveBrokers.size())).id);
+        }
+    }
+
+    private static List<BrokerMetadata> getAliveBrokers(MetadataImage image) {
```

Comment:
Do we really need to use `BrokerMetadata`? It seems we can remove `getAliveBrokers` and then simplify `getRandomAliveBroker`.
```java
    private static Optional<Integer> getRandomAliveBroker(MetadataImage image) {
        var aliveBrokers = image.cluster().brokers().values().stream()
                .filter(broker -> !broker.fenced())
                .map(BrokerRegistration::id)
                .toList();
        if (aliveBrokers.isEmpty()) return Optional.empty();
        return Optional.of(aliveBrokers.get(ThreadLocalRandom.current().nextInt(aliveBrokers.size())));
    }
```

Reply:
Update it. Thanks.